Additional edits to Roxygen vignette

ArtemSokolov opened the pull request:
Hi Dirk,

I think I understand your earlier comment about `@export` keyword. The default `NAMESPACE` file generated by `Rcpp.package.skeleton()` exports everything. Since `roxygen2` won't modify the file generated by another source, `@export` keywords have no effect on function visibility. For educational purposes, I added a paragraph that suggests creating a backup of the default `NAMESPACE` and letting roxygen2 generate its own. The user can then compare the two and decide if they want to expose everything (default `Rcpp.package.skeleton()` behavior) or have finer control over individual functions via the `@export` keyword.

I also made another pass through the instructions to check reproducibility. Based on this, I made the following edits:

1. Reformatted the source to 80-character width, as requested.
2. Minor rephrasing in third paragraph of Introduction.
3. Added `roxygen2` to the `install.packages` code block.
4. Corrected the usage of `Rcpp::Rcpp.package.skeleton()`
5. Suggested the user modify `src/rcpp_hello_world.cpp` generated by `Rcpp.package.skeleton()`. Added the matching suggestion to remove the dangling `man/*.Rd` files after `src/rcpp_hello_world.cpp` is modified.

Let me know what you think. If you would prefer that we have the user write a new `src/mult.cpp` alongside the existing `src/rcpp_hello_world.cpp` instead of overwriting the latter, I can make the appropriate edits.

Best,
-Artem

ArtemSokolov committed:
Additional edits to Roxygen vignette

eddelbuettel commented:
> The default NAMESPACE file generated by Rcpp.package.skeleton() exports everything. 

Yes. That is my default working mode.  I edit NAMESPACE by hand, zero dependencies etc pp.  So the skeleton() works the same way.  Let me take a look...

eddelbuettel commented:
Briefly:

1.  Reformatted.  I already did.  I used 100 cols because on my current setup at home that is roughly half a screen.  So now you just created new diffs for no real gain.  I am sorry if I gave the impression that 80 cols is hard fast rule. I believe I said 80 or 90 or 100.  

2. Will try to isolate that.

3. Upon reflection I think I am against this here if we use `Rcpp.package.skeleton()`.  The base R workflow does not involve `roxygen2`, see _e.g._ comments by Duncan Murdoch on the lists.  Now, it is a useful and popular package so we may mention it as an alternative (and write that other piece about alternate development modes).

4. Oops. Will try to isolate that too.

5. Will try to isolate that.  The helloWorld.Rd is indeed weird and may need clean-up.

ArtemSokolov commented:
Re: 3. Without `install.packages("roxygen2")`, the following has the potential to throw an error:

https://github.com/RcppCore/rcpp-gallery/blob/284b44a1e2d6c114b5893c4a6d6f5c3dbea0878b/src/2020-01-02-documenting-rcpp-packages.Rmd#L163

eddelbuettel commented:
I clearly missed that.  And I also may have missed that the post is about roxygen :)  So yes, assume it is installed.  

How do you feel about focusing this post on _e.g._ `roxygenize(".", roclets="rd")` and leaving `NAMESPACE` and all that out?

ArtemSokolov commented:
Yes, that's fine. I guess keeping things as simple as possible with minimal actions required from the reader is probably the way to go.

I have to focus on something else for a while, but I will make another pass later today (or tomorrow morning). I'll revert back to your original formatting to keep the new changes more visible in the diff. (I've dealt with enough 80-character sticklers in my career, that I just go along with it nowadays ;) But you're right: your original comment was 80 or 100.)

eddelbuettel commented:
Sorry about the extra work and really appreciate that you are still working with the stickler typing this :)

(And I sometimes say 'issue tickets before pr' to has things out ...)

ArtemSokolov committed:
Simplified NAMESPACE by letting Rcpp.package.skeleton() handle it

ArtemSokolov committed:
Updated render

ArtemSokolov commented:
OK, latest commits to this PR should reflect everything we talked about. I reverted back to the original formatting and simplified the handling of NAMESPACE.

The "Files changed" tab should be much easier to parse now. ;)

ArtemSokolov commented on _posts/2020-01-02-documenting-rcpp-packages.md:
```diff
@@ -25,40 +25,36 @@ and have it appear in the package reference manual.
 
 
 {% highlight r %}
-#' @useDynLib testpkg
```
Not needed since skeleton() already places `useDynLib()` in `NAMESPACE`. This particular `roxygen` directive will have no effect.

eddelbuettel replied:
Agreed.

ArtemSokolov commented on _posts/2020-01-02-documenting-rcpp-packages.md:
```diff
@@ -84,24 +80,14 @@ which Rcpp will translate into `#'` in the corresponding `RcppExports.R` file:
 double mult( double v1, double v2 ) {return v1*v2;}
 {% endhighlight %}
 
-You may notice two `export` keywords used in the comment block. The first, `@export`, tells Roxygen2
```
Similarly to the `useDynLib` point above, `@export` directives are not needed, since we're leaving `NAMESPACE` as is. I removed all such directives and the corresponding English description.

eddelbuettel replied:
Yep.

eddelbuettel commented:
Thanks again for the very careful fixes.  I just added one small things -- `royxgenize()` needs a `roclets="rd"` argument.  While at it, I closed/summarized with an outlook to other tools and why we kept it simple.  Let me know if it is good enough, else we pass another round.

ArtemSokolov commented:
Looks great, Dirk! I think we can call it done.

eddelbuettel commented:
Thanks for this.  And I think you convinved this grumpy old fellow to add a vignette about different workflows.  While Rcpp is ten years old and I still work the same way with it. not everybody else does so maybe we should talk about it.

Do you want to take the lead on that?  And no rush...

ArtemSokolov commented:
I think that, to do a proper job, we will probably want to solicit input from the community. I suspect that my own view of possible workflows is fairly limited: I usually start package development by copying DESCRIPTION from another project, modifying it by hand and then using `devtools` to automatically generate everything else. I also do most of my work in Emacs + ESS, so my experience with RStudio is relatively limited.

All that to say is that my lack of breadth may be a possible limiting factor. In line with your earlier 'issue tickets before pr', perhaps the best way forward is to open an issue and invite long-term Rcpp users to describe their preferred workflows. The vignette can then systematically categorize and review the various approaches that the community uses.

eddelbuettel commented:
I am usually a bit fearful of the unspecified / amorphous "community" and rather work with clearly tasks and collaborators.  

But I see where you are coming from, and I will mull over whether a vignette would be better.  I still like it as a simpler, less formal, Gallery post though.
